File: Dynamic Programming/Maximum_sum_of_non_adjacent_elements.cpp
#include<iostream>
#include<vector>
using namespace std;

int solveRecursive(vector<int> nums,int n){
    if ( n < 0)
        return 0;
    if (n == 0)
        return nums[0];

    int incl = solveRecursive(nums, n - 2) + nums[n];
    int execu = solveRecursive(nums , n - 1);

    return max(incl,execu);
}

int solveMeMorization(vector<int> nums,int n,vector<int> &dp){
    if ( n < 0)
        return 0;
    if (n == 0)
        return nums[0];
    
    if (dp[n] != -1)
        return dp[n];

    int incl = solveMeMorization(nums, n - 2 , dp) + nums[n];
    int execu = solveMeMorization(nums , n - 1 , dp);

    dp[n] = max(incl,execu);
    return dp[n];
}

int maximumNonAdjacentSum(vector<int> &nums){
    // Write your code here.
    int n = nums.size();

    // return solveRecursive(nums,n-1);

    vector<int> dp(n,-1);
    return solveMeMorization(nums,n-1,dp);

}
int main(){
    vector<int> nums = {2, 1, 4, 9};
    cout<<maximumNonAdjacentSum(nums);
}

File: Dynamic Programming/Maximum_sum_of_non_adjacent_elements.java
import java.util.ArrayList;
import java.util.Arrays;

public class Maximum_sum_of_non_adjacent_elements {
    public static int maximumNonAdjacentSum(ArrayList<Integer> nums) {

        // Solving using dp
        int prev1 = nums.get(0);
        int prev2 = 0;

        for (int i = 1; i < nums.size(); i++) {
            int incl = nums.get(i) + prev2;
            int excl = prev1;

            int ans = Math.max(incl, excl);
            prev2 = prev1;
            prev1 = ans;
        }

        return prev1;
    }

    public static void main(String[] args) {
        ArrayList<Integer> nums = new ArrayList<Integer>(Arrays.asList(2, 1, 4, 9));

        System.out.println(maximumNonAdjacentSum(nums));
    }
}